ambry-router/src/main/java/com/github/ambry/router/ReplicateBlobCallback.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright 2022 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.clustermap.DataNodeId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.Callback;
+
+
+/**
+ * If some operation like {@link TtlUpdateOperation} fails due to some replicas don't have the Blob.
+ * We use {@link ReplicateBlobOperation} to replicate the blob to the NOT_FOUND replicas.
+ * After that, retry the failed operation for example the {@link TtlUpdateOperation}
+ * ReplicateBlobCallback is the callback function of the {@link ReplicateBlobOperation},
+ * also it stores the retry state.
+ */
+public class ReplicateBlobCallback implements Callback<Void> {
+
+  public enum RetryState {
+    REPLICATING_BLOB, // replicating the blob
+    REPLICATION_DONE, // blob replication is done
+    RETRYING          // retry the original failed operation
+  }
```

Comment:
One boolean "isDone" is not enough. It can be used to indicate if the ReplicateBlob is done. But we also have retry state(retrying the TtlUpdate).
In checkAndMaybeComplete, we couldn't differentiate these two cases
1. first TtlUpdate failed, isDone is true
2. second TtlUpdate failed. isDone is true

For case 1, we should retry. 
For case 2, we should stop retrying and fail the request. 
Comparing with two boolean, isDone, hasRetried, I prefer using one single enum state.